src/bun.js/bindings/bindings.cpp

```diff
@@ -272,11 +309,74 @@ AsymmetricMatcherResult matchAsymmetricMatcher(JSGlobalObject* globalObject, JSC
+            if (otherProp.isObject()) {
+                ThrowScope scope = DECLARE_THROW_SCOPE(globalObject->vm());
+                if (Bun__deepMatch<true>(otherProp, patternObject, globalObject, &scope, false, true)) {
+                    return AsymmetricMatcherResult::PASS;
+                }
+            }
+        }
+
+        return AsymmetricMatcherResult::FAIL;
+    } else if (auto* expectCloseTo = jsDynamicCast<JSExpectCloseTo*>(matcherPropCell)) {
+        if (!readFlagsAndProcessPromise(matcherProp, expectCloseTo, flags, globalObject, otherProp))
+            return AsymmetricMatcherResult::FAIL;
+
+        if (!otherProp.isNumber()) {
+            // disable the "not" flag here, because if not a number it should still return FAIL when negated
+            flags = flags & ~FLAG_NOT;
+            return AsymmetricMatcherResult::FAIL;
+        }
+
+        JSValue expectedValue = expectCloseTo->m_numberValue.get();
+        JSValue digitsValue = expectCloseTo->m_digitsValue.get();
+
+        double received = otherProp.toNumber(globalObject);
+        double expected = expectedValue.toNumber(globalObject);
+
+        const double infinity = std::numeric_limits<double>::infinity();
+
+        // special handing because (Infinity - Infinity) or (-Infinity - -Infinity) is NaN
+        if ((received == infinity && expected == infinity) || (received == -infinity && expected == -infinity)) {
+            return AsymmetricMatcherResult::PASS;
+        } else {
+            int32_t digits = digitsValue.toInt32(globalObject);
+
+            double threshold = 0.5 * std::pow(10.0, -digits);
+            bool isClose = std::abs(expected - received) < threshold;
+            return isClose ? AsymmetricMatcherResult::PASS : AsymmetricMatcherResult::FAIL;
+        }
+    } else if (auto* customMatcher = jsDynamicCast<JSExpectCustomAsymmetricMatcher*>(matcherPropCell)) {
+        if (!readFlagsAndProcessPromise(matcherProp, customMatcher, flags, globalObject, otherProp))
+            return AsymmetricMatcherResult::FAIL;
+
+        // ignore the "not" flag here, because the custom matchers handle it themselves (accessing this.isNot)
+        // and it would result in a double negation
+        flags = flags & ~FLAG_NOT;
+
+        bool passed = ExpectCustomAsymmetricMatcher__execute(customMatcher->wrapped(), JSValue::encode(matcherProp), globalObject, JSValue::encode(otherProp));
+        return passed ? AsymmetricMatcherResult::PASS : AsymmetricMatcherResult::FAIL;
     }
 
     return AsymmetricMatcherResult::NOT_MATCHER;
 }
 
+AsymmetricMatcherResult matchAsymmetricMatcher(JSGlobalObject* globalObject, JSValue matcherProp, JSValue otherProp, ThrowScope* throwScope)
+{
+    ExpectFlags flags = ExpectFlags();
+    AsymmetricMatcherResult result = matchAsymmetricMatcherAndGetFlags(globalObject, matcherProp, otherProp, throwScope, flags);
+    if (result != AsymmetricMatcherResult::NOT_MATCHER && (flags & FLAG_NOT)) {
```

Comment:
In jest both `expect.not.any` and `expect.not.anything` are not valid asymmetric matchers. We will need to throw for both of them.
<img width="605" alt="Screenshot 2023-11-28 at 4 14 24 PM" src="https://github.com/oven-sh/bun/assets/35280289/aa0c2f62-bc3d-4d79-8d16-702869a22706">

Reply:
I know, but I feel like we can support them "for free" as a feature, as it's not gonna break any existing jest test. They might be useful, e.g.:
- match a null or undefined with `expect.not.anything()`
- match something that's not a Date with `expect.not.any(Date)`

Reply:
I agree, let's keep them. `expect.not.any(Class)` seems very useful